cocotb/triggers.py

```diff
@@ -831,9 +805,9 @@ def with_timeout(trigger, timeout_time, timeout_unit=None):
 
     .. versionadded:: 1.3
     """
-    timeout_timer = cocotb.triggers.Timer(timeout_time, timeout_unit)
-    res = yield [timeout_timer, trigger]
-    if res is timeout_timer:
-        raise cocotb.result.SimTimeoutError
-    else:
-        return res
+
+    async def timeout():
+        await Timer(timeout_time, timeout_unit)
+        raise cocotb.result.SimTimeoutError()
+
+    return await First(timeout(), trigger)
```

Comment:
Any reason for the rewrite here?

Reply:
In the previous implementation, we used `yield []` which returns the trigger than fires. Now we have to use `First` which returns the result of the first trigger that fires. This makes it hard to distinguish which trigger fired. The implementation was going to have to change, so I went with an implementation that tests new functionality.

Reply:
`yield [a, b]` is identical to `yield First(a, b)`, which is almost identical to `await First(a, b)`.

Both return the _result_ of the first trigger that fires, not the trigger itself.

However, it happens that the result of a `Timer` _is_ the trigger itself.

Reply:
Returned to original implementation.